docs(dbt-cloud): add guide for setting up dbt Cloud with Dagster (#10708)
dagster-io/dagster: docs/content/integrations/dbt-cloud.mdx

height={673}
/>

---

## Step 1: Connecting to dbt Cloud

> The first step in using dbt Cloud with Dagster is to tell Dagster how to connect to your dbt Cloud instance using a dbt Cloud [resource](/concepts/resources). This resource contains information on where the dbt Cloud instance is located and any credentials sourced from environment variables that needed to access it.

Reviewer: ```suggestion
The first step in using dbt Cloud with Dagster is to tell Dagster how to connect to your dbt Cloud instance using a dbt Cloud [resource](/concepts/resources). This resource contains information on where the dbt Cloud instance is located and any credentials sourced from environment variables that are needed to access it.
```

from dagster_dbt import dbt_cloud_resource

dbt_cloud_instance = dbt_cloud_resource.configured(
    {
        "auth_token": {"env": "DBT_CLOUD_API_TOKEN"},
        "account_id": {"env": "DBT_CLOUD_ACCOUNT_ID"},
    }
)
```

## Step 2: Loading dbt Cloud models as assets

In this step, you'll load the dbt Cloud models managed by a dbt Cloud job into Dagster as assets. For context, a dbt Cloud job defines set of commands to run for a dbt Cloud project. The dbt Cloud models managed by a dbt Cloud job are the models that are run by the job after filtering options are respected.

Using our dbt Cloud resource, we can retrieve information about the models that the dbt Cloud job is managing.

```python startafter=start_load_assets_from_dbt_cloud_job endbefore=end_load_assets_from_dbt_cloud_job file=/integrations/dbt/dbt_cloud.py dedent=4
from dagster_dbt import load_assets_from_dbt_cloud_job

# Use the dbt_cloud_instance resource we defined in Step 1, and the job_id from Prerequisites
dbt_cloud_assets = load_assets_from_dbt_cloud_job(
    dbt_cloud=dbt_cloud_instance,
    job_id=33333,
)
```

<Note>

>   We assume that your dbt Cloud job only has one command: one of either `dbt

Reviewer: ```suggestion
  We currently require that your dbt Cloud job only has one command: one of either `dbt
```

  run` or `dbt build`, along with any optional command line arguments.
</Note>

The <PyObject module="dagster_dbt" object="load_assets_from_dbt_cloud_job" /> function loads the dbt Cloud models into Dagster as assets, creating one Dagster asset for each model.

When invoked, the function:

1. Invokes your dbt Cloud job with command overrides to compile your dbt project,
2. Parses the metadata provided by dbt Cloud, and
3. Generates a set of software-defined assets reflecting the models in the project managed by the dbt Cloud job. Materializing these assets will run the dbt Cloud job that is represented by the loaded assets.

## Step 3: Schedule dbt Cloud job runs

> Now that your dbt Cloud assets, you can define a Dagster job that materializes some or all of these assets, triggering the underlying dbt Cloud job.

Reviewer: ```suggestion
Now that your dbt Cloud assets are loaded, you can define a Dagster job that materializes some or all of these assets, triggering the underlying dbt Cloud job.
```

> Instead of guessing when dbt Cloud should run, you can schedule them based on their upstream or downstream dependencies, external events using a sensor, or cron schedule.

Reviewer: ```suggestion
Instead of guessing when dbt Cloud should run, you can schedule assets based on their upstream or downstream dependencies, external events using a sensor, or a cron schedule.
```

height={673}
/>

---

## Step 1: Connecting to dbt Cloud

The first step in using dbt Cloud with Dagster is to tell Dagster how to connect to your dbt Cloud instance using an dbt Cloud [resource](/concepts/resources). This resource contains information on where the dbt Cloud instance is located and any credentials needed to access it.

```python startafter=start_define_dbt_cloud_instance endbefore=end_define_dbt_cloud_instance file=/integrations/dbt/dbt_cloud.py dedent=4
from dagster_dbt import dbt_cloud_resource

> dbt_cloud_instance = dbt_cloud_resource.configured(

Reviewer: We should tell people where they get these in dbt Cloud or provide a link

    {
        "auth_token": {"env": "DBT_CLOUD_API_TOKEN"},
        "account_id": {"env": "DBT_CLOUD_ACCOUNT_ID"},
    }
)
```

## Step 2: Loading dbt Cloud models as assets

> In this step, you'll load the dbt Cloud models managed by a dbt Cloud job into Dagster as assets. For context, a dbt Cloud job defines set of commands to run for a dbt Cloud project.

Reviewer: I think its worth clarifying what "managed by a dbt Cloud job" means. For instance, are all the models imported in the dbt job only "manages" a subset?

Using our dbt Cloud resource, we can retrieve information about the models that the dbt Cloud job is managing.

```python startafter=start_load_assets_from_dbt_cloud_job endbefore=end_load_assets_from_dbt_cloud_job file=/integrations/dbt/dbt_cloud.py dedent=4
from dagster_dbt import load_assets_from_dbt_cloud_job

# Use the dbt_cloud_instance resource we defined in Step 1, and the job_id from Prerequisites
dbt_cloud_assets = load_assets_from_dbt_cloud_job(
    dbt_cloud=dbt_cloud_instance,

>     job_id=33333,

Reviewer: We should tell people where to get the job id, in my use of dbt cloud it wasn't obvious that I had to get it from the URL

)
```

The <PyObject module="dagster_dbt" object="load_assets_from_dbt_cloud_job" /> function loads the dbt Cloud models into Dagster as assets, creating one Dagster asset for each model.

When invoked, the function:

1. Invokes your dbt Cloud job with command overrides to compile your dbt project,
2. Parses the metadata provided by dbt Cloud, and

> 3. Generates a set of software-defined assets reflecting the models in the project. These assets share the same underlying op, which will invoke dbt Cloud to run the dbt Cloud job represented by the loaded assets.

Reviewer: ```suggestion
3. Generates a set of software-defined assets reflecting the models in the project managed by the dbt Cloud job. Materializing these assets will invoke dbt Cloud to run the dbt Cloud job represented by the loaded assets.
```

Reviewer: My suggestion is mostly a reflection of confusion, I don't understand what "these assets share the same underlying op" implies for the use of these assets. 

I think it'd also be worth adding a little more color about what happens when assets are materialized, e.g. does the dbt  Cloud job run verbatim, or do we only execute `dbt run`? If the dbt Cloud job includes tests, does asset materialization?

> ## Step 3: Schedule dbt Cloud job runs

Reviewer: It'd be great to highlight the "so what?" of the integration here.

"Instead of guessing when dbt should run, now that your dbt Cloud assets are available in Dagster you can schedule them based on their up or downstream dependencies, based on external events using a sensor, or ... "
